### ISSUE TITLE
The Device lifecycle: absent, not ready yet, and lost

### ISSUE THREAD

**Issue by dvoyni** [dvoyni/cog#485]
## Parent

[audio: a plugin whose sounds are commanded, positioned, and mixed in pure Go](https://github.com/dvoyni/cog/issues/294). Specs: `slots/sound/docs/specs/sound.md`, `extensions/otosound/docs/specs/otosound.md`.

## What to build

A web game can draw its *click to enable sound* prompt, unplugging headphones is a non-event, and a machine with no sound card still runs the game.

**Audio never stops a game from starting.** Registration always succeeds and `Startup` neither blocks nor fails — `gfx`'s shape one Slot over, where the driver supplies its Backend before the device exists. A Device that can never be opened is reported **once, by the Adapter**, after which it behaves exactly as `nosound` does.

The Device is absent, not ready yet, or lost, and a game sees one thing in all three. One rule covers them: **a Voice's playhead advances whether or not anyone can hear it.** That is deliberately unlike the Clip rule, where a Voice waiting on its Clip starts at the head — a Clip load is a bounded hiccup the engine is actively fixing, and Device readiness may never come at all.

## Acceptance criteria

- [ ] `sound.Device` is a read-only resource carrying `Ready`, `Name`, `SampleRate`, `Channels` and `Latency`
- [ ] **`sound` never reports a not-ready backend** — no counterpart to `gfx.ErrBackendNotReady`, because on web not-ready is normal and only the Adapter knows which cause it is
- [ ] `otosound.ErrDeviceUnavailable` is reported once, by the Adapter, when the open fails; the game runs
- [ ] A Voice recorded while the Device is not ready advances silently and is simply mid-Clip when the Device arrives, or has already ended with its event delivered on schedule
- [ ] **Device loss is reported never**: `Ready` goes false, playback keeps being simulated, and the Adapter retries the open once a second indefinitely
- [ ] On recovery Voices resume where the world is *now*, and the restart burst — a decoder open and a seek per streamed Voice — does not land on the device thread
- [ ] A second `Engine` shares the process's `oto` context and takes its own player; the first composition's `Config` wins, reported once as `otosound.ErrDeviceConfigIgnored`, and `Device` reports the **actual** values in force
- [ ] `nosound` reports `Ready` true and `Latency` zero, both true rather than placeholders, so a prompt gated on `Ready` does not hang under it
- [ ] A paused engine does not change `Device`
- [ ] `Device()` is a field read at the seam, polled once per flush, with nothing pushed
- [ ] The **Gap** is recorded in the implementation: `oto` v3's behaviour on a device change was never measured, so whether the retry loop ever fires is open

## Blocked by

- #477 — `otosound` makes the first sound

**Comment by dvoyni** [comment]
## Two things `otosound` left for this ticket, and one measurement nobody has taken

[otosound makes the first sound](https://github.com/dvoyni/cog/issues/477) landed in `e2fec7b` and built the device far enough to play, which is where its device handling stops being guesswork and starts being this ticket's.

**1. oto's device-change behaviour is still a Gap in the spec, and it is now blocking something real.** A 1 Hz poll of `ctx.Err()` / `player.Err()` with an indefinite once-a-second retry is implemented, but it has **never been tested against an actual device switch** — the spike never measured it. The failure mode to check for: if `oto` simply keeps writing into a dead device and never surfaces an error, the retry never fires and the game goes permanently silent with `Device.Ready` still true. If that is what happens, polling is the wrong mechanism and an explicit close-and-reopen is needed. **Measure it on real hardware** — unplug headphones, switch the default output device, sleep and wake the machine — rather than reasoning about it.

**2. A lost device retains destroys forever.** Released Clips are held until the Mixer's atomic applied-counter passes the batch that destroyed them, which is what keeps *the Mixer never frees* true. Two cases exist and only one is handled:

- *No Mixer has ever pulled* — freed at once, which is what makes `otosound` behave as `nosound` under a device that never opened. Handled.
- *Pulled once, then lost* — the counter stops, and every subsequent release is held for as long as the device stays gone. **Not handled.** Bounded by what the game releases after the loss, so it is a leak rather than a crash, but it is unbounded in time.

**3. Recovery is not a burst.** The Mixer's voice table survives a reattach, so live Voices continue from their own playheads rather than resyncing to where the world is now. That is a deliberate non-decision in this slice, not an oversight — the spec's burst description is about the streamed tier's decoder reopens, which is [#482](https://github.com/dvoyni/cog/issues/482). Whether a Voice that played through a 30-second device outage should resume mid-clip or be re-pinned to the world's current state is this ticket's to say.

**Comment by dvoyni** [comment]
## Built and landed — one criterion outstanding, and it needs hardware

Landed as `9e8ba32` on `main`. **This ticket stays open** for the one item no agent can discharge: `oto` v3's behaviour on a real device change has still never been measured, which is the spec's own recorded Gap. Everything else is met and verified.

### What now works

**Recovery re-pins every live Voice to where the world is *now*, and it is the Slot's job.** The flush reads `backend.Device()` *before* it collects the batch, compares `Ready` with what the resource held last tick, and on a false→true transition restates every live, non-pending Voice as a `VoiceStart` at its current playhead. A start is the only operation at the seam that carries a position, so **a restart is the resync**; it carries that tick's `Params` too, so nothing is owed beside it. Bounded by `MaxVoices`, paid once per arrival, never per tick.

It cannot be the Adapter's, because an Adapter has no playheads at all — `sound` computes them. An Adapter left to resync itself could only resume its own mixer table where that table stopped, which is where the world *was*.

**The same transition covers all three states.** A Device that was never ready becoming ready is a web game's first click, and gets the identical restart.

**A lost Device no longer holds every release forever** — the leak flagged when [otosound](https://github.com/dvoyni/cog/issues/477) landed. `ring.detach()` puts `reclaim` back on its *never pulled* branch, so releases free at once again. The ordering is the contract and is commented as one: `Ready` goes false → the player stops → the ring detaches.

Stopping the player was `oto`'s `Pause()`, and **that was wrong** — oto documents `Pause` as *continuing to read the source* so playback resumes gapless, and a source still being read is a device thread that still exists. It is now `PauseAndStopReading()`, documented as blocking until an ongoing read finishes and starting no other, which is the barrier a detach needs. A correctness fix, not a rename.

**Loss is now unreportable however it presents itself.** A Device lost after a successful open whose *reattach* then failed used to call `fail()` and be reported as `ErrDeviceUnavailable` — a loss reported, which the spec forbids. An `everReady` flag makes `fail()` return early once a Device has been ready even once, so only a Device that could never be opened at all is reported, still once.

### Acceptance criteria

- [x] `sound.Device` is read-only, carrying `Ready`, `Name`, `SampleRate`, `Channels`, `Latency`
- [x] **`sound` never reports a not-ready backend** — no error path in the Slot reads `Ready` at all; `TestSoundNeverReportsANotReadyDevice` runs 16 ticks with a Voice playing under a Device that never opens and fails if the kernel reports anything
- [x] `otosound.ErrDeviceUnavailable` reported once by the Adapter; the game runs — asserted by waiting for the retry to run *again* and checking the same `*deviceFailure` pointer is still the recorded one
- [x] A Voice recorded while not ready advances silently and is mid-Clip on arrival, or has already ended on schedule — playhead 0.5 after 32 silent ticks, arrival tick emits one start at exactly 33/64 s; and `ReasonFinished` on tick 32 with no Device ever
- [x] **Loss reported never**: `Ready` goes false, playback keeps being simulated, the open is retried once a second indefinitely
- [x] On recovery, Voices resume where the world is now — 4 ticks, 30 ticks of outage, recovery start at 35/64 s, and **no second restart** on the tick after
- [x] …and the restart burst does not land on the device thread — **for the tier that exists**. The resident restart is an index, and the structural device-thread test still holds. The streamed tier's decoder-open-and-seek per Voice is [#482](https://github.com/dvoyni/cog/issues/482), as this ticket's own text says
- [x] A second Engine shares the context, first `Config` wins, reported once, `Device` reports the **actual** values
- [x] `nosound` reports `Ready` true and `Latency` zero, both true rather than placeholders
- [x] A paused engine does not change `Device` — **at the level the engine currently expresses pause.** Engine-level `app` pause is not wired into `sound` at all yet; only per-Voice `Params.Paused` exists, and that subscription is [#481](https://github.com/dvoyni/cog/issues/481)'s. Nothing in `otosound`'s device path reads pause in any form, and `Device` is written only by the device goroutine
- [x] The **Gap** is recorded in the implementation — a doc block on `backend.watch` naming the failure mode and the run that would settle it

### How it was verified

`go build`, `go vet`, `GOOS=js GOARCH=wasm go build` clean, and a web build composing `otosound` still correctly refuses to compile. `go test ./...` green across the module on the merged tree; the four audio packages at `-count=10`.

Each new behaviour was **checked red before green**, with the failures recorded: resync disabled → *the Device arrived and 0 Voices were started* (×3 tests); `ring.detach()` commented out → *timed out waiting for the ring to be detached*; the `everReady` guard disabled → *a Device that came back reported the headphones came out on the way*.

**`go test -race` was unavailable** — `runtime/cgo: cgo.exe: exit status 2`. Everything crossing a goroutine boundary ran at `-count=10` instead; race-freedom is **unverified**, not established.

The merge with [Buses](https://github.com/dvoyni/cog/issues/478) was checked semantically, not just textually: the resync arm emits `slot.voiceParams()`, which now scales by `audibility()`, so **a Bus volume survives a recovery** rather than being restored at unity.

## To close this ticket — the measurement, for a human

On a machine with a real sound card:

1. Compose a game with `otosoundplugin` and play a long looping Clip.
2. Print `Device().Ready` once a second (a System reading `kernel.Read[*sound.Device]` is enough).
3. One at a time, with the Clip audibly playing: **unplug the headphones**; **switch the default output device** in the OS mixer; **sleep the machine and wake it**.
4. For each: does `Ready` go false within a second or two, and does sound come back within a second or two of the device returning?

**A run where the audio stops but `Ready` stays true is the failure mode the spec names** — `oto` writing into a dead device and never surfacing an error. That answer changes the mechanism: polling would be wrong, and an explicit close-and-reopen on a cadence would be needed. The shape around it does not change either way, so only `backend.watch` and `attach`'s loss branch would need rework.

## What a later ticket must settle

1. **The staging batch grows without bound during a long outage.** Once the ring's eight slots are full, every further tick merges into the staging batch, whose `ops` slice appends past its `3×MaxVoices` capacity, and each `opStart` pins a `*clipData`. This is a second unbounded-in-time growth of the same family this ticket fixed, but the ticket named only the pending-destroy list, so scope was not widened. **Someone should decide whether a detached ring also resets its staging batch** — the argument against is that it would drop starts the seam promises never to lose, so the honest fix is probably a cap plus a stated loss rule. That is a design question, not an implementation one.
2. **`VoiceStart.Loop` is never set by `VoicesCollect`**, by the ordinary start path or the new resync path. Looping is not wired above the seam yet. **If [#481](https://github.com/dvoyni/cog/issues/481) adds it to the normal start, the resync arm needs the same field**, or a recovery will silently un-loop every looping Voice.
3. **A restart is a hard cut, not a ramp** — `opStart` sets `cur = target`, which is right here since nothing was audible during the outage. But `PauseAndStopReading` keeps its buffered data and `Play` resumes gapless, so up to one buffer (10 ms) of pre-loss audio may replay at the moment of recovery. Below the *nothing is promised below the block* line, so it was left.
4. **`retryEvery` became a `cadence` field on `backend`.** The spec rejects a configurable retry cadence as a knob, and this is not one: no `Config` reaches it, `newBackend` seeds it from the constant, and only the suite writes it. Without the seam, the loss and recovery tests take ~2 s each and ~22 s under `-count=10`.

### Spec edits

Both `sound.md` and `otosound.md` gained text here — **58 lines added, none removed**: how *where the world is now* is said and whose job it is, that it covers all three states, the release-freeing on loss, the `PauseAndStopReading` ordering, loss being unreportable however it presents, and a Gap note pointing at the code that records it.

**Comment by dvoyni** [comment]
## The Gap is answered for Windows, from `oto`'s source rather than from a hardware run

`oto` v3.5.0's Windows driver was read. **The failure mode this spec feared does not occur there**, and the mechanism built here is a backstop rather than the primary recovery.

### What `oto` does on Windows

`driver_windows.go` tries **WASAPI**, falls back to **WinMM**, and installs a **null context** if neither finds a device.

On the WASAPI render thread, `driver_wasapi_windows.go:447` polls the default endpoint **every 500 ms**:

```go
if now := time.Now(); now.Sub(last) >= 500*time.Millisecond {
    switched, err := c.isDeviceSwitched()   // GetDefaultAudioEndPoint(eRender, eConsole)
```

A switch yields `errDeviceSwitched`. That error — together with `AUDCLNT_E_DEVICE_INVALIDATED`, `AUDCLNT_E_RESOURCES_INVALIDATED` and `RPC_E_DISCONNECTED` — is explicitly **not** joined into `c.err` (`:248`). It goes to `restart()`, which reinitialises on the new default with 50 ms → 500 ms backoff, up to **24 attempts**, *draining the mux at real time* so playback resumes near its current position rather than replaying a backlog.

**So on a default-device switch or an unplug:**

- audio **follows** to the new device in roughly half a second;
- `ctx.Err()` stays nil, so **`Device().Ready` stays true**;
- otosound's 1 Hz poll **never fires**, because there is nothing left to recover.

`c.err` is set — and otosound's loop finally takes over — only after 24 consecutive transient failures (~10 s of no usable device) or on a non-transient error.

### What this settles

**The spec's recorded Gap** — *"if `oto` simply keeps writing into a dead device and never errors, the retry never fires and an explicit close-and-reopen is needed"* — **is contradicted on Windows/WASAPI.** `oto` both detects and recovers. The mechanism built here is correct as a backstop and should be described as one.

**A hazard named when this landed is narrower than stated.** *A Device lost after the Mixer has pulled holds every release forever* needs the applied-counter to stop advancing. `oto` keeps draining the mux throughout its restart, so the counter keeps moving and releases keep being freed. That hazard only bites once `oto` has given up entirely, which is a far smaller window than the note implied.

**Scope of the claim.** This is a source reading of `oto` **v3.5.0**, **Windows only**, and no hardware run was performed. macOS (`driver_darwin.go`) and Linux (ALSA/PulseAudio) are different drivers and are unexamined. The analysis is pinned to that version; a dependency bump should re-check it.

### The test plan, revised

Recorded for whoever does run it, because the expected outcome is not what this ticket originally assumed:

1. **Switch the default output device** — expect audio to *follow* and `Ready` to stay **true**. This exercises `oto`, not otosound. It is still worth doing: it is what a real player does, and it disproves the feared mode.
2. **Unplug the headphones that are currently default** — same, via the invalidation path.
3. **Sleep and wake** — exercises `oto`'s `suspended` branch, which resets its retry budget on resume.
4. **Disable every output device for ~15 s, then re-enable** — the *only* one of the four that reaches otosound's code, because it is the only one that exhausts `oto`'s 24 attempts. Expect `Ready` false, the game still running, the playhead still advancing, nothing reported, and sound back within a second of re-enabling.

### A defect found while reading, filed separately

`oto`'s Windows fallback chain ends in a **null context** whose `Err()` returns `nil` and whose loop drains the mux forever. otosound sets `Ready: true` whenever the open returns no error, so **on a Windows machine with no audio device at all, `ErrDeviceUnavailable` is never reported and `Device().Ready` is true** while nothing can be heard.

That contradicts this ticket's third criterion, so it is **not** being closed over: it is filed as its own ticket, because the fix belongs to otosound's open path rather than to the lifecycle this ticket built.

Closing on the strength of the analysis above, at the maintainer's direction.
